core/rmutex.c

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright (C) 2016 Theobroma Systems Design & Consulting GmbH
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    core_sync Synchronization
+ * @brief       Recursive Mutex for thread synchronization
+ * @{
+ *
+ * @file
+ * @brief       RIOT synchronization API
+ *
+ * @author      Martin Elshuber <martin.elshuber@theobroma-systems.com>
+ *
+ * The recursive mutex implementation is inspired by the implementetaion of
+ * Nick v. IJzendoorn <nijzendoorn@engineering-spirit.nl>
+ * @see https://github.com/RIOT-OS/RIOT/pull/4529/files#diff-8f48e1b9ed7a0a48d0c686a87cc5084eR35
+ *
+ */
+
+#include <stdio.h>
+#include <inttypes.h>
+
+#include "rmutex.h"
+#include "thread.h"
+#include "assert.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+void rmutex_lock(rmutex_t *rmutex)
+{
+    kernel_pid_t owner;
+
+    /* try to lock the mutex */
+    DEBUG("rmutex %" PRIi16" : trylock\n", thread_getpid());
+    if (mutex_trylock(&rmutex->mutex) == 0) {
+        DEBUG("rmutex %" PRIi16" : mutex already held\n", thread_getpid());
+        /* mutex is already held
+         *
+         * Case 1: the current thread holds the mutex
+         *     Invariant 1: holds
```

Comment:
Invariant is an incorrect word i think.

the point is that the outcome of the condition 
if ( owner != thread_getpid() ) 
well never change even if rmutex->owner is changed by an other thread (holder of the mutex)
i will update the comment occrdingly